core/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -192,13 +199,49 @@ static boolean execute(SearchContext searchContext, final IndexSearcher searcher
 
             boolean timeoutSet = scrollContext == null && searchContext.timeout() != null &&
                 searchContext.timeout().equals(SearchService.NO_TIMEOUT) == false;
+
+            final Runnable timeoutRunnable;
             if (timeoutSet) {
-                // TODO: change to use our own counter that uses the scheduler in ThreadPool
-                // throws TimeLimitingCollector.TimeExceededException when timeout has reached
-                collectors.add(createTimeoutCollectorContext(searchContext.timeEstimateCounter(), searchContext.timeout().millis()));
+                final Counter counter = searchContext.timeEstimateCounter();
+                final long startTime = counter.get();
+                final long timeout = searchContext.timeout().millis();
+                final long maxTime = startTime + timeout;
+                timeoutRunnable = () -> {
+                    final long time = counter.get();
+                    if (time > maxTime) {
+                        queryResult.searchTimedOut(true);
+                        throw new CollectionTerminatedException();
```

Comment:
Do you think it is necessary? The number of segments should be bounded so checking all of them should not be much more costly than stopping for all of them at once, at allows to keep things a bit simpler?

I only did things this way for cancellation so that we still check on a per-segment basis of low-level cancellation is disabled.